front: fix added exception with rs change group not displayed

SharglutDev opened the pull request:
- Fix the rolling stock image not displayed for added exceptions.
- Use the rolling stock list from the scenario in `useOccurrences` to do the matching instead of an api call
- Remove the rs api call for indexed occurrences in favor of this new list
- Rename `useOccurrences` with .ts extension since it doesn't return any JSX

fix #12164

SharglutDev committed:
front: fix added exception with rs change group

The matching was done for index occurrences but not added ones.
Pass down the rolling stock list provided by the scenario to the
occurrence builder to match with the added exception's rolling stock.

Signed-off-by: SharglutDev <p.filimon75@gmail.com>

emersion commented:
Ow, doing HTTP requests in a loop like this for each occurrence is not great, and the rtk-query cache isn't used since queries aren't subscribed so each update might cause new HTTP requests to be sent…

We have a list of rolling stocks at hand in the timetable, could we drop all of the async stuff and use that instead?

SharglutDev committed:
front: simplify rolling stock indexed exceptions

Use the rolling stock list provided by the scenario to match the one
from the exception. This avoids an unnecessary api call.

Signed-off-by: SharglutDev <p.filimon75@gmail.com>

SharglutDev committed:
front: rename useOccurrences

As this hook doesn't return any JSX, the file doesn't need to be a .tsx
but only a .ts

Signed-off-by: SharglutDev <p.filimon75@gmail.com>

emersion commented on front/src/modules/trainschedule/components/Timetable/PacedTrain/hooks/useOccurrences.tsx:
```diff
@@ -109,17 +112,13 @@ const useOccurrences = (pacedTrain: PacedTrainWithDetails) => {
-          occurrenceRollingStock = await promisedRollingStock.unwrap();
-          // we don't want to subscribe to the endpoint to prevent unnecessary calls
-          promisedRollingStock.unsubscribe();
+        if (exception.rolling_stock && rollingStockList) {
```
This is inside a `useEffect()`, so I think we need to add this to the dependencies list?

SharglutDev replied:
Good catch. Done

emersion commented on front/src/modules/trainschedule/components/Timetable/PacedTrain/hooks/useOccurrences.tsx:
```diff
@@ -66,13 +63,9 @@ const useOccurrences = (
-          const promisedRollingStock = getRollingStockByName({
-            rollingStockName: correspondingException.rolling_stock.rolling_stock_name,
-          });
-          occurrenceRollingStock = await promisedRollingStock.unwrap();
```
With this change, we no longer need this to be a `useEffect()`, we can turn it back into a `useMemo()`. Do you want to keep this for a follow-up PR, or do you want to do it in this PR?

SharglutDev replied:
Oh yes, you're right ! Forgot I was the one who added it for this rolling stock fetching. I'll add a commit.

SharglutDev committed:
front: remove useEffect from useOccurrences

Since we no longer make any api call in the hook, the logic is not async
anymore so we don't need the useEffect and useState, we can switch back to a
useMemo.

Signed-off-by: SharglutDev <p.filimon75@gmail.com>

emersion reviewed (approved):
LGTM!

Reading this PR commit by commit was very comfortable, thanks for taking the time to do this!

SharglutDev commented:
> LGTM!
> 
> Reading this PR commit by commit was very comfortable, thanks for taking the time to do this!

Learned from the best 😎
